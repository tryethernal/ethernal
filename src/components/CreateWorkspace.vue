<template>
    <v-card elevation="0">
        <v-card-text v-if="(!user.plan || user.plan == 'free') && user.onboarded">
            <v-alert dense text type="error">Free plan users are limited to one workspace. <a href="#" @click.stop="goToBilling()">Upgrade</a> to our Premium plan to create more.</v-alert>
        </v-card-text>
        <v-card-text v-else>
            <v-alert v-show="errorMessage" dense text type="error" v-html="errorMessage"></v-alert>
            <div class="mb-2">
                <v-alert type="warning" class="my-2" v-if="isUsingSafari">
                    Safari is preventing websites from making CORS requests to localhost. This will prevent you from connecting to a local blockchain. If you want to do so, you'll need to use another browser.
                    If you want to connect to a remote chain, or are not using Safari, you can ignore this message.
                </v-alert>
                <a id="detectServers" href="#" @click.prevent="detectNetwork()">Detect Networks</a>&nbsp;
                <v-tooltip top>
                    <template v-slot:activator="{ on }">
                        <v-icon small v-on="on">mdi-help-circle-outline</v-icon>
                    </template>
                    This will send a RPC request asking for a network ID to 127.0.0.1 on http and ws protocols on commonly used ports (7545, 8545 and 9545).<br>The address will be displayed below if the request is successful.
                </v-tooltip>
                <ul v-show="detectedNetworks.length">
                    <li v-for="(address, idx) in detectedNetworks" :key="idx">
                        {{ address }}&nbsp;<a href="#" :id="`serverDetected-${idx}`" @click.prevent="rpcServer = address">Use</a>
                    </li>
                </ul>
                <div v-show="noNetworks">
                    No networks detected. If you were expecting something, make sure they are running on 7545, 8545 or 9545 and that your browser is not blocking requests to localhost (looking at you Brave & Safari ðŸ‘€!).
                </div>
            </div>
            <v-form @submit.prevent="createWorkspace(name, rpcServer)" v-model="valid">
                <v-text-field
                    :rules="[v => !!v || 'Name is required']"
                    outlined v-model="name" id="workspaceName" label="Name*" placeholder="My Ethereum Project" hide-details="auto" class="mb-2" required></v-text-field>
                <v-text-field
                    :rules="[
                        v => this.isUrlValid(v) || 'RPC needs to be a valid URL',
                        v => !!v || 'RPC server is required'
                    ]"
                    outlined v-model="rpcServer" id="workspaceServer" label="RPC Server*" placeholder="ws://localhost:8545" hide-details="auto" class="mb-2" required></v-text-field>
                <v-select outlined required label="Chain" v-model="chain" :items="availableChains" hide-details="auto"></v-select>

                <v-card-actions>
                    <v-spacer></v-spacer>
                    <v-btn id="createWorkspace" :loading="loading" color="primary" :disabled="!valid" type="submit">Create</v-btn>
                </v-card-actions>
            </v-form>
        </v-card-text>
    </v-card>
</template>
<script>
const ipaddr = require('ipaddr.js');
import { mapGetters } from 'vuex';
export default {
    name: 'CreateWorkspace',
    data: () => ({
        availableChains: [],
        chain: 'ethereum',
        errorMessage: null,
        loading: false,
        name: null,
        rpcServer: null,
        localNetwork: false,
        detectedNetworks: [],
        noNetworks: false,
        workspace: null,
        valid: false
    }),
    mounted() {
        this.availableChains = Object.values(this.chains).map((chain) => ({ text: chain.name, value: chain.slug }));
    },
    methods: {
        async createWorkspace(name, rpcServer) {
            this.loading = true;
            try {
                this.workspace = await this.server.initRpcServer(rpcServer);
            } catch(error) {
                this.loading = false;
                if (error.message.indexOf('Invalid URL') > -1)
                    return this.errorMessage = `
                        URL of the rpc server looks invalid. Make sure you entered a valid ws(s) or http(s) url.
                    `;
                else if (this.localNetwork && (rpcServer.startsWith('http://') || rpcServer.startsWith('ws://'))) {
                    return this.errorMessage = `
                        Can't connect to rpc server, 
                        make sure your node is up, and reachable from the browser.<br>
                        If you are using Brave, you'll need to disable Shields. 
                        Try disabling adblockers as well.<br>
                        Another option is to setup a public URL such as <a href="https://ngrok.com/" target="_blank">ngrok</a>, and use https to connect.
                    `;
                }
                else if (!this.localNetwork && (rpcServer.startsWith('http://') || rpcServer.startsWith('ws://'))) {
                    return this.errorMessage = `
                        Can't connect to remote rpc server.<br>
                        Make sure your node is up, supports "eth_chainId" & "net_version" requests, and that you've allowed app.tryethernal.com to 
                        <a href="https://experienceleague.adobe.com/docs/target/using/experiences/vec/troubleshoot-composer/mixed-content.html" target="_blank">load insecure content</a>.<br>
                        Try using a secure connection (https or wss) if possible.
                    `;
                }
                else if (!this.localNetwork && rpcServer.startsWith('wss://')) {
                    return this.errorMessage = `
                        Can't connect to remote rpc server. 
                        Make sure your node is up, and supports "eth_chainId" & "net_version" requests. 
                        Try using https if possible.
                    `;
                }
                else
                    return this.errorMessage = `
                        Can't connect to remote rpc server.<br>
                        Make sure your node is up, and supports "eth_chainId" & "net_version" requests.
                    `;
            }

            this.server.createWorkspace(name, { ...this.workspace, chain: this.chain })
                .then(({ data }) => this.$emit('workspaceCreated', data))
                .catch(error => {
                    if (error.response && error.response.data)
                        return this.errorMessage = error.response.data;
                    else
                        return this.errorMessage = 'Error while creating workspace';
                })
                .finally(() => this.loading = false)
        },
        detectNetwork() {
            this.noNetworks = false;
            this.server.searchForLocalChains().then((res) => {
                this.detectedNetworks = res;
                if (!res.length) {
                    this.noNetworks = true;
                }
            });
        },
        isUrlValid(url) {
            try {
                new URL(url);
                return true;
            } catch(error) {
                return false;
            }
        },
        goToBilling() {
            this.$emit('goToBilling');
        }
    },
    computed: {
        ...mapGetters([
            'user',
            'chains'
        ]),
        isUsingSafari() {
            return navigator.vendor.match(/apple/i) && !navigator.userAgent.match(/crios/i) && !navigator.userAgent.match(/fxios/i);
        },
    },
    watch: {
        rpcServer() {
            try {
                if (!this.isUrlValid(this.rpcServer)) {
                    return;
                }
                const hostname = new URL(this.rpcServer).hostname;
                const localStrings = ['private', 'linkLocal', 'loopback', 'carrierGradeNat', 'localhost'];
                if (hostname == 'localhost') {
                    this.localNetwork = true;
                }
                else {
                    this.localNetwork = ipaddr.isValid(hostname) && localStrings.indexOf(ipaddr.parse(hostname).range()) > -1;
                }
            } catch(error) {
                console.log(error);
            }
        }
    }
}
</script>
