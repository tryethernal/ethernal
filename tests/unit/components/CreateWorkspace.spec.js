import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';

import CreateWorkspace from '@/components/CreateWorkspace.vue';

beforeEach(() => jest.clearAllMocks());

describe('CreateWorkspace.vue', () => {
    let helper;

    beforeEach(() => {
        helper = new MockHelper();
    });

    it('Should display a warning message if user has already been onboarded & on a free plan', () => {
        const wrapper = helper.mountFn(CreateWorkspace, {
            getters: {
                user: jest.fn().mockReturnValue({
                    plan: 'free',
                    onboarded: true
                })
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should not display a warning message if user has already been onboarded & on a premium plan', () => {
        const wrapper = helper.mountFn(CreateWorkspace, {
            getters: {
                user: jest.fn().mockReturnValue({
                    plan: 'premium',
                    onboard: true
                })
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should let the user create a new workspace', async () => {
        const name = 'Hardhat';
        const rpcServer = 'http://127.0.0.1:8545';

        jest.spyOn(helper.mocks.server, 'initRpcServer').mockResolvedValue(true);
        jest.spyOn(helper.mocks.server, 'createWorkspace').mockResolvedValue(true);

        const wrapper = helper.mountFn(CreateWorkspace, {
            attachTo: document.body
        });

        await wrapper.setData({ name, rpcServer });
        await flushPromises();

        await wrapper.find('#createWorkspace').trigger('click');
        await flushPromises();

        expect(wrapper.emitted().workspaceCreated).toBeTruthy();
        wrapper.destroy();
    });

    it('Should display an error if it cannot connect to the chain', async () => {
        const name = 'Hardhat';
        const rpcServer = 'http://127.0.0.1';

        jest.spyOn(helper.mocks.server, 'initRpcServer').mockRejectedValue(new Error('Error'));

        const wrapper = helper.mountFn(CreateWorkspace, {
            attachTo: document.body
        });

        await wrapper.setData({ name, rpcServer });
        await flushPromises();

        await wrapper.find('#createWorkspace').trigger('click');
        await flushPromises();

        expect(wrapper.emitted().workspaceCreated).toBeFalsy();
        expect(wrapper.html()).toMatchSnapshot();
        wrapper.destroy();
    });

    it('Should display a warning if browser is Safari', () => {
        jest.spyOn(navigator, 'vendor', 'get').mockReturnValue('apple');
        const wrapper = helper.mountFn(CreateWorkspace);

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display locally running networks', async () => {
        const wrapper = helper.mountFn(CreateWorkspace);
        await wrapper.find('#detectServers').trigger('click');
        await wrapper.vm.$nextTick();
        await wrapper.find('#serverDetected-0').trigger('click');

        expect(wrapper.vm.rpcServer).toBe('http://127.0.0.1:8545');
        expect(wrapper.html()).toMatchSnapshot();
    });
});
