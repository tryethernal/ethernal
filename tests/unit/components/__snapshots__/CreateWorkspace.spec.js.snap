// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateWorkspace.vue Should display a warning if browser is Safari 1`] = `
<div class="v-card v-sheet theme--light elevation-0">
  <div class="v-card__text">
    <div role="alert" class="v-alert v-sheet theme--dark v-alert--dense v-alert--text error--text" style="display: none;"></div>
    <div class="mb-2">
      <div role="alert" class="v-alert my-2 v-sheet theme--dark warning">
        <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-exclamation theme--dark"></i>
          <div class="v-alert__content">
            Safari is preventing websites from making CORS requests to localhost. This will prevent you from connecting to a local blockchain. If you want to do so, you'll need to use another browser.
            If you want to connect to a remote chain, or are not using Safari, you can ignore this message.
          </div>
        </div>
      </div> <a id="detectServers" href="#">Detect Networks</a>&nbsp;
      <span class="v-tooltip v-tooltip--top"><!----><i aria-hidden="true" class="v-icon notranslate mdi mdi-help-circle-outline theme--light" style="font-size: 16px;"></i></span>
      <ul style="display: none;"></ul>
      <div style="display: none;">
        No networks detected. If you were expecting something, make sure they are running on 7545, 8545 or 9545 and that your browser is not blocking requests to localhost (looking at you Brave &amp; Safari ðŸ‘€!).
      </div>
    </div>
    <form novalidate="novalidate" class="v-form">
      <div class="v-input mb-2 v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-text-field--placeholder">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="workspaceName" class="v-label theme--light" style="left: 0px; position: absolute;">Name*</label><input required="required" id="workspaceName" type="text"></div>
          </div>
        </div>
      </div>
      <div class="v-input mb-2 v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-text-field--placeholder">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="workspaceServer" class="v-label theme--light" style="left: 0px; position: absolute;">RPC Server*</label><input required="required" id="workspaceServer" type="text"></div>
          </div>
        </div>
      </div>
      <div class="v-input v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-85" class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-select__slot"><label for="input-85" class="v-label theme--light" style="left: 0px; position: absolute;">Chain</label>
              <div class="v-select__selections"><input required="required" id="input-85" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden" value="ethereum">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class="v-card__actions">
        <div class="spacer"></div> <button type="submit" disabled="disabled" class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default" id="createWorkspace"><span class="v-btn__content">Create</span></button>
      </div>
    </form>
  </div>
</div>
`;

exports[`CreateWorkspace.vue Should display a warning message if user has already been onboarded & on a free plan 1`] = `
<div class="v-card v-sheet theme--light elevation-0">
  <div class="v-card__text">
    <div role="alert" class="v-alert v-sheet theme--dark v-alert--dense v-alert--text error--text">
      <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark error--text"></i>
        <div class="v-alert__content">Free plan users are limited to one workspace. <a href="#">Upgrade</a> to our Premium plan to create more.</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CreateWorkspace.vue Should display an error if it cannot connect to the chain 1`] = `
<div class="v-card v-sheet theme--light elevation-0">
  <div class="v-card__text">
    <div role="alert" class="v-alert v-sheet theme--dark v-alert--dense v-alert--text error--text" style="">
      Can't connect to rpc server,
      make sure your node is up, and reachable from the browser.<br>
      If you are using Brave, you'll need to disable Shields.
      Try disabling adblockers as well.<br>
      Another option is to setup a public URL such as <a href="https://ngrok.com/" target="_blank">ngrok</a>, and use https to connect.
    </div>
    <div class="mb-2">
      <div role="alert" class="v-alert my-2 v-sheet theme--dark warning">
        <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-exclamation theme--dark"></i>
          <div class="v-alert__content">
            Safari is preventing websites from making CORS requests to localhost. This will prevent you from connecting to a local blockchain. If you want to do so, you'll need to use another browser.
            If you want to connect to a remote chain, or are not using Safari, you can ignore this message.
          </div>
        </div>
      </div> <a id="detectServers" href="#">Detect Networks</a>&nbsp;
      <span class="v-tooltip v-tooltip--top"><!----></span><i aria-hidden="true" class="v-icon notranslate mdi mdi-help-circle-outline theme--light" style="font-size: 16px;"></i>
      <ul style="display: none;"></ul>
      <div style="display: none;">
        No networks detected. If you were expecting something, make sure they are running on 7545, 8545 or 9545 and that your browser is not blocking requests to localhost (looking at you Brave &amp; Safari ðŸ‘€!).
      </div>
    </div>
    <form novalidate="novalidate" class="v-form">
      <div class="v-input mb-2 v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined v-text-field--placeholder">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: -24px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="workspaceName" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Name*</label><input required="required" id="workspaceName" type="text"></div>
          </div>
        </div>
      </div>
      <div class="v-input mb-2 v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined v-text-field--placeholder">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: -24px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="workspaceServer" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">RPC Server*</label><input required="required" id="workspaceServer" type="text"></div>
          </div>
        </div>
      </div>
      <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-62" class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: -24px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-select__slot"><label for="input-62" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Chain</label>
              <div class="v-select__selections">
                <div class="v-select__selection v-select__selection--comma">Ethereum</div><input required="required" id="input-62" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden" value="ethereum">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class="v-card__actions">
        <div class="spacer"></div> <button type="submit" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary" id="createWorkspace"><span class="v-btn__content">Create</span></button>
      </div>
    </form>
  </div>
</div>
`;

exports[`CreateWorkspace.vue Should display locally running networks 1`] = `
<div class="v-card v-sheet theme--light elevation-0">
  <div class="v-card__text">
    <div role="alert" class="v-alert v-sheet theme--dark v-alert--dense v-alert--text error--text" style="display: none;"></div>
    <div class="mb-2">
      <div role="alert" class="v-alert my-2 v-sheet theme--dark warning">
        <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-exclamation theme--dark"></i>
          <div class="v-alert__content">
            Safari is preventing websites from making CORS requests to localhost. This will prevent you from connecting to a local blockchain. If you want to do so, you'll need to use another browser.
            If you want to connect to a remote chain, or are not using Safari, you can ignore this message.
          </div>
        </div>
      </div> <a id="detectServers" href="#">Detect Networks</a>&nbsp;
      <span class="v-tooltip v-tooltip--top"><!----></span><i aria-hidden="true" class="v-icon notranslate mdi mdi-help-circle-outline theme--light" style="font-size: 16px;"></i>
      <ul style="">
        <li>
          http://127.0.0.1:8545&nbsp;<a href="#" id="serverDetected-0">Use</a></li>
      </ul>
      <div style="display: none;">
        No networks detected. If you were expecting something, make sure they are running on 7545, 8545 or 9545 and that your browser is not blocking requests to localhost (looking at you Brave &amp; Safari ðŸ‘€!).
      </div>
    </div>
    <form novalidate="novalidate" class="v-form">
      <div class="v-input mb-2 v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-text-field--placeholder">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="workspaceName" class="v-label theme--light" style="left: 0px; position: absolute;">Name*</label><input required="required" id="workspaceName" type="text"></div>
          </div>
        </div>
      </div>
      <div class="v-input mb-2 v-input--has-state v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-text-field--placeholder error--text">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: -24px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="workspaceServer" class="v-label v-label--active theme--light error--text" style="left: 0px; position: absolute;">RPC Server*</label><input required="required" id="workspaceServer" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light error--text" role="alert">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                <div class="v-messages__message">RPC needs to be a valid URL</div>
              </transition-group-stub>
            </div>
          </div>
        </div>
      </div>
      <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-103" class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: -24px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-select__slot"><label for="input-103" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Chain</label>
              <div class="v-select__selections">
                <div class="v-select__selection v-select__selection--comma">Ethereum</div><input required="required" id="input-103" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden" value="ethereum">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class="v-card__actions">
        <div class="spacer"></div> <button type="submit" disabled="disabled" class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default" id="createWorkspace"><span class="v-btn__content">Create</span></button>
      </div>
    </form>
  </div>
</div>
`;

exports[`CreateWorkspace.vue Should not display a warning message if user has already been onboarded & on a premium plan 1`] = `
<div class="v-card v-sheet theme--light elevation-0">
  <div class="v-card__text">
    <div role="alert" class="v-alert v-sheet theme--dark v-alert--dense v-alert--text error--text" style="display: none;"></div>
    <div class="mb-2">
      <div role="alert" class="v-alert my-2 v-sheet theme--dark warning">
        <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-exclamation theme--dark"></i>
          <div class="v-alert__content">
            Safari is preventing websites from making CORS requests to localhost. This will prevent you from connecting to a local blockchain. If you want to do so, you'll need to use another browser.
            If you want to connect to a remote chain, or are not using Safari, you can ignore this message.
          </div>
        </div>
      </div> <a id="detectServers" href="#">Detect Networks</a>&nbsp;
      <span class="v-tooltip v-tooltip--top"><!----><i aria-hidden="true" class="v-icon notranslate mdi mdi-help-circle-outline theme--light" style="font-size: 16px;"></i></span>
      <ul style="display: none;"></ul>
      <div style="display: none;">
        No networks detected. If you were expecting something, make sure they are running on 7545, 8545 or 9545 and that your browser is not blocking requests to localhost (looking at you Brave &amp; Safari ðŸ‘€!).
      </div>
    </div>
    <form novalidate="novalidate" class="v-form">
      <div class="v-input mb-2 v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-text-field--placeholder">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="workspaceName" class="v-label theme--light" style="left: 0px; position: absolute;">Name*</label><input required="required" id="workspaceName" type="text"></div>
          </div>
        </div>
      </div>
      <div class="v-input mb-2 v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-text-field--placeholder">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="workspaceServer" class="v-label theme--light" style="left: 0px; position: absolute;">RPC Server*</label><input required="required" id="workspaceServer" type="text"></div>
          </div>
        </div>
      </div>
      <div class="v-input v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-21" class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span class="notranslate">â€‹</span></legend>
            </fieldset>
            <div class="v-select__slot"><label for="input-21" class="v-label theme--light" style="left: 0px; position: absolute;">Chain</label>
              <div class="v-select__selections"><input required="required" id="input-21" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden" value="ethereum">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class="v-card__actions">
        <div class="spacer"></div> <button type="submit" disabled="disabled" class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default" id="createWorkspace"><span class="v-btn__content">Create</span></button>
      </div>
    </form>
  </div>
</div>
`;
